### PR TITLE
8322758: Eliminate -Wparentheses warnings in C2 code

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to eliminate some -Wparentheses warnings.  In most
cases, this involved simply adding a few parentheses to make some implicit
operator precedence explicit.

In PhaseIdealLoop::rc_predicate, I also added a comment describing the test
being performed, since it didn't seem obvious even with the additional
parentheses.

Testing: mach5 tier1

Also ran mach5 tier1 with these changes in conjunction enabling -Wparentheses
and other changes needed to make that work.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322758](https://bugs.openjdk.org/browse/JDK-8322758): Eliminate -Wparentheses warnings in C2 code (**Enhancement** - P4)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to [8acc005e](https://git.openjdk.org/jdk/pull/17199/files/8acc005e9da69161185a6fda10d8ef2342eb2fb0)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [8acc005e](https://git.openjdk.org/jdk/pull/17199/files/8acc005e9da69161185a6fda10d8ef2342eb2fb0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17199/head:pull/17199` \
`$ git checkout pull/17199`

Update a local copy of the PR: \
`$ git checkout pull/17199` \
`$ git pull https://git.openjdk.org/jdk.git pull/17199/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17199`

View PR using the GUI difftool: \
`$ git pr show -t 17199`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17199.diff">https://git.openjdk.org/jdk/pull/17199.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17199#issuecomment-1871676202)

### CODE REVIEW

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 1]
```diff
@@ -797,7 +797,8 @@ BoolNode* PhaseIdealLoop::rc_predicate(IdealLoopTree* loop, Node* ctrl, int scal
   overflow = false;
   Node* max_idx_expr = nullptr;
   const TypeInt* idx_type = TypeInt::INT;
-  if ((stride > 0) == (scale > 0) == upper) {
+  // same signs and upper, or different signs and not upper.
+  if (((stride > 0) == (scale > 0)) == upper) {
```
Comment (theRealAph) [inline src/hotspot/share/opto/loopPredicate.cpp:801]: This is rather l33t code, but I guess it's OK with the comment. This
```suggestion
  _Bool same_signs = (stride > 0) == (scale > 0);
  if ((same_signs & upper)
      || (!same_signs && !upper)) {
```
generates slightly more code with GCC -O2.  I'd be happy with either.

**Review by theRealAph — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/opto/loopPredicate.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/opto/loopPredicate.cpp:801]: I agree it's a little odd, but I don't feel strongly about it, so leaving it as is.

**Comment by kimbarrett** [comment]
Thanks for reviews @theRealAph and @vnkozlov .

**Pull request closed without merging**